server/service/service_osquery.go

```diff
@@ -446,15 +447,92 @@ func (svc *Service) detailQueriesForHost(ctx context.Context, host *fleet.Host)
 }
 
 func (svc *Service) shouldUpdate(lastUpdated time.Time, interval time.Duration, hostID uint) bool {
-	var jitter time.Duration
-	if svc.config.Osquery.MaxJitterPercent > 0 {
-		maxJitter := int64(svc.config.Osquery.MaxJitterPercent) * int64(interval) / 100.0
-		jitter = time.Duration((int64(hostID) + svc.getJitterSeed()) % maxJitter)
+	svc.jitterMu.Lock()
+	defer svc.jitterMu.Unlock()
+
+	if svc.jitterH[interval] == nil {
+		svc.jitterH[interval] = &jitterHashTable{
+			maxCapacity: 1,
+			bucketCount: int(int64(svc.config.Osquery.MaxJitterPercent) * int64(interval) / 100.0),
+			buckets:     make(map[int]int),
+			cache:       make(map[uint]time.Duration),
+		}
 	}
+
+	jitter := svc.jitterH[interval].jitterForHost(hostID)
+	//jitter := jitterForHost(
+	//	svc.config.Osquery.MaxJitterPercent,
+	//	svc.getJitterSeed(),
+	//	interval,
+	//	hostID,
+	//)
 	cutoff := svc.clock.Now().Add(-(interval + jitter))
 	return lastUpdated.Before(cutoff)
 }
 
+type jitterHashTable struct {
+	mu          sync.Mutex
+	maxCapacity int
+	bucketCount int
+	buckets     map[int]int
+	cache       map[uint]time.Duration
+}
+
+func (jh *jitterHashTable) jitterForHost(hostID uint) time.Duration {
+	jh.mu.Lock()
+	if jitter, ok := jh.cache[hostID]; ok {
+		jh.mu.Unlock()
+		return jitter
+	}
+
+	possibleBucket := int(hostID) % jh.bucketCount
+	if jh.buckets[possibleBucket] < jh.maxCapacity {
+		jh.buckets[possibleBucket]++
+		jitter := time.Duration(possibleBucket) * time.Minute
+		jh.cache[hostID] = jitter
+
+		jh.mu.Unlock()
+		return jitter
+	}
+
+	lastChecked := possibleBucket
+	for {
+		possibleBucket = (possibleBucket + 1) % jh.bucketCount
+
+		// if we went around already, there's no more capacity
+		if possibleBucket == lastChecked {
+			break
+		}
+
+		// if the next bucket has capacity, great!
+		if jh.buckets[possibleBucket] < jh.maxCapacity {
+			jh.buckets[possibleBucket]++
+			jitter := time.Duration(possibleBucket) * time.Minute
+			jh.cache[hostID] = jitter
```

Comment:
Do we need to invalidate the cache eventually?

Reply:
Only if we want to save RAM, we can use a LRU cache.

Reply:
Ya, I realized this won't change like the other approach where we periodically generated a new random seed.